crates/driver/src/infra/solver/protocol_fees.rs

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    pub fn apply_to_order(&self, order: &competition::Order, weth: eth::WethAddress) -> Order {
+        // In case of volume based fees, fee withheld by driver might be higher than the
+        // surplus of the solution. This would lead to violating limit prices when
+        // driver tries to withhold the volume based fee. To avoid this, we artificially
+        // adjust the order limit amounts (make then worse) before sending to solvers,
+        // to force solvers to only submit solutions with enough surplus to cover the
+        // fee.
+        //
+        // https://github.com/cowprotocol/services/issues/2440
+        let mut available = order.available(weth);
+        if self.driver_must_apply_protocol_fee(order.signature.signer.into()) {
+            if let Some(fees::FeePolicy::Volume { factor }) = order.protocol_fees.first() {
+                match order.side {
+                    Side::Buy => {
+                        // reduce sell amount by factor
+                        available.sell.amount = available
+                            .sell
+                            .amount
+                            .apply_factor(1.0 / (1.0 + factor))
+                            .unwrap_or_default();
+                    }
+                    Side::Sell => {
+                        // increase buy amount by factor
+                        available.buy.amount = available
+                            .buy
+                            .amount
+                            .apply_factor(1.0 / (1.0 - factor))
+                            .unwrap_or_default();
+                    }
+                }
+            }
+        }
+        Order {
+            uid: order.uid.into(),
+            sell_token: available.sell.token.into(),
+            buy_token: available.buy.token.into(),
+            sell_amount: available.sell.amount.into(),
+            buy_amount: available.buy.amount.into(),
+            fee_amount: available.user_fee.into(),
+            kind: order.side.into(),
+            partially_fillable: order.is_partial(),
+            class: order.kind.into(),
+            fee_policies: self.protocol_fee_to_solvers(weth.into(), order.protocol_fees.as_ref()),
+        }
+    }
+
+    pub fn apply_to_solution(&self, solution: Solution) -> Result<Solution, Error> {
+        // Apply protocol fees only if the drivers is set to handler the fees
+        if self.fee_handler != FeeHandler::Driver {
+            return Ok(solution);
+        }
+
+        let mut trades = Vec::with_capacity(solution.trades.len());
+        for trade in solution.trades {
+            match &trade {
+                Trade::Fulfillment(fulfillment) => match fulfillment.order().kind {
+                    order::Kind::Market | order::Kind::Limit { .. } => {
+                        // Apply protocol fees only if the drivers is set to handler the fees
+                        if self.driver_must_apply_protocol_fee(
```

Comment:
If the `Driver` is set to calculate the protocol fees, we need to check each order in case the trader address is whitelisted. This should be really cheap since the searching time is O(1).